lighthouse-core/audits/time-to-interactive.js

```diff
@@ -0,0 +1,164 @@
+   */
+  static audit(artifacts) {
+    // We start looking at Math.Max(FMPMetric, visProgress[0.85])
+    return FMPMetric.audit(artifacts).then(fmpResult => {
+      if (fmpResult.value === -1) {
+        return generateError(fmpResult.debugString);
+      }
+      const fmpTiming = parseFloat(fmpResult.rawValue);
+      const timings = fmpResult.extendedInfo.timings;
+
+      // Process the trace
+      const tracingProcessor = new TracingProcessor();
+      const model = tracingProcessor.init(artifacts.traceContents);
+      const endOfTraceTime = model.bounds.max;
+
+      // TODO: Wait for DOMContentLoadedEndEvent
+      // TODO: Wait for UA loading indicator to be done
+
+      // look at speedline results for 85% starting at FMP
+      const visualProgress = artifacts.Speedline.frames.map(frame => {
+        return {
+          progress: frame.getProgress(),
+          time: frame.getTimeStamp()
+        };
+      });
+      const fMPts = timings.fMPfull + timings.navStart;
+      const visuallyReady = visualProgress.find(frame => {
+        return frame.time >= fMPts && frame.progress >= 85;
+      });
+      const visuallyReadyTiming = visuallyReady.time - timings.navStart;
+
+      // Find first 500ms window where Est Input Latency is <50ms at the 90% percentile.
+      let startTime = Math.max(fmpTiming, visuallyReadyTiming) - 50;
+      let endTime;
+      let currentLatency = Infinity;
+      const percentile = 0.9;
+      const threshold = 50;
+      let foundLatencies = [];
+
+      // When we've found a latency that's good enough, we're good.
+      while (currentLatency > threshold) {
+        // While latency is too high, increment just 50ms and look again.
+        startTime += 50;
+        endTime = startTime + 500;
+        // If there's no more room in the trace to look, we're done.
+        // TODO return an error instead
+        if (endTime > endOfTraceTime) {
+          return;
+        }
+        // Get our expected latency for the time window
+        const latencies = TracingProcessor.getRiskToResponsiveness(
+          model, artifacts.traceContents, startTime, endTime, [percentile]);
+        const estLatency = latencies[0].time.toFixed(2);
+        foundLatencies.push(Object.assign({}, {
+          startTime: startTime.toFixed(1),
+          estLatency
+        }));
+        // console.log('At', startTime.toFixed(2), '90 percentile est latency is ~', estLatency);
+        // Grab this latency and try the threshold again
+        currentLatency = estLatency;
+      }
+      const timeToInteractive = startTime.toFixed(1);
+
+      // Use the CDF of a log-normal distribution for scoring.
+      //   < 1200ms: score≈100
+      //   5000ms: score=50
+      //   >= 15000ms: score≈0
+      const distribution = TracingProcessor.getLogNormalDistribution(SCORING_MEDIAN,
+          SCORING_POINT_OF_DIMINISHING_RETURNS);
+      let score = 100 * distribution.computeComplementaryPercentile(timeToInteractive);
```

Comment:
subtraction is obviously easy to do, but it might be worth eventually having it be explicit. As we add critical path profiling stuff this may come out naturally, though.